server/core/auth.js

```diff
@@ -173,26 +173,28 @@ module.exports = {
       user.groups.forEach(grp => {
         const grpId = _.isObject(grp) ? _.get(grp, 'id', 0) : grp
         _.get(WIKI.auth.groups, `${grpId}.pageRules`, []).forEach(rule => {
+          let hasRole = _.intersection(rule.roles, checkPermissions).length > 0
+
           switch (rule.match) {
             case 'START':
-              if (_.startsWith(`/${page.path}`, `/${rule.path}`)) {
+              if (_.startsWith(`/${page.path}`, `/${rule.path}`) && hasRole) {
```

Comment:
Move `hasRole` at the beginning of the expression, so it's faster to exit the condition if false. ( repeat for other rules below )

Reply:
I got rid of hasRole all together and wrapped the entire switch in an if statement that checks for the role in the permission with _.intersection.